Add FeeCurrencyDirectory to Anvil migrations

arthurgousset opened the pull request:
### Description

Adds the new `FeeCurrencyDirectory` contract to the Anvil migration script.

### Changes

1. deploys and proxies the `FeeCurrencyDirectory` contract
2. registers `FeeCurrencyDirectory` in the `Registry`
3. adds three existing cStables (cUSD, cEUR, and cREAL) to the `FeeCurrencyDirectory`

### Other changes

1. adds `is IOracle` to the class definition in `SortedOracles`
4. removes a duplicate import in `Migrations.s.sol`

### Tested

Yes, tested locally:
- [x] get the `FeeCurrencyDirectory` address from the `Registry`
- [x] get registered currencies from the `FeeCurrencyDirectory` contract
- [x] get the exchange rate for a currency from the `FeeCurrencyDirectory` contract

<details>
  <summary>Steps to reproduce tests locally</summary>

Requirement:
1. Have foundry installed
2. Be in the `packages/protocol` directory

Tests:

1. Run migration script (in a different Terminal)

    ```sh
    $ ./migrations_sol/create_and_migrate_anvil_devchain.sh
    ```

1. Check that Anvil is running

    ```sh
    $ nc -z localhost 8546
    Connection to localhost port 8546 [tcp/*] succeeded!
    ```
    
    Source: [start_anvil.sh](https://github.com/celo-org/celo-monorepo/blob/a36eae5bc2c9c2071fb942199b1ba265f3937c17/packages/protocol/migrations_sol/start_anvil.sh#L7)

    Anvil is running ✅

1. get the `FeeCurrencyDirectory` address from the `Registry`:

    ```sh
    export REGISTRY_ADDRESS="0x000000000000000000000000000000000000ce10"
    export ANVIL_PORT=8546
    cast call \
    $REGISTRY_ADDRESS \
    "getAddressForStringOrDie(string calldata identifier)" \
    "FeeCurrencyDirectory" \
    --rpc-url=http://127.0.0.1:$ANVIL_PORT
    
    0x00000000000000000000000042fe5a2a61ed9705eb2f08a04a58ceb606d22f6a
    ```
    
    ```sh
    cast abi-decode \
    "getAddressForStringOrDie(string calldata identifier)(address)" \
    "0x00000000000000000000000042fe5a2a61ed9705eb2f08a04a58ceb606d22f6a"
    
    0x42Fe5a2A61ed9705eb2F08a04A58CEB606D22f6a
    ```

    `FeeCurrencyDirectory` is registered in the `Registry` ✅

1. get registered currencies from the `FeeCurrencyDirectory` contract

    ```sh
    export FEECURRENCYDIRECTORY_PROXY_ADDRESS=0x42Fe5a2A61ed9705eb2F08a04A58CEB606D22f6a
    export ANVIL_PORT=8546
    
    cast call \
    $FEECURRENCYDIRECTORY_PROXY_ADDRESS \
    "getCurrencies()" \
    --rpc-url=http://127.0.0.1:$ANVIL_PORT
    
    0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003000000000000000000000000e6774be4e5f97db10cafb4c00c74cfbdcdc434d9000000000000000000000000b7a33b4ad2b1f6b0a944232f5c71798d27ad92720000000000000000000000002a3733dbc31980f02b12135c809b5da33bf3a1e9
    ```
    
    ```sh
    cast abi-decode \
    "getCurrencies()(address[] memory)" \
    "0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003000000000000000000000000e6774be4e5f97db10cafb4c00c74cfbdcdc434d9000000000000000000000000b7a33b4ad2b1f6b0a944232f5c71798d27ad92720000000000000000000000002a3733dbc31980f02b12135c809b5da33bf3a1e9"
    
    [0xe6774BE4E5f97dB10cAFB4c00C74cFbdCDc434D9, 0xb7a33b4ad2B1f6b0a944232F5c71798d27Ad9272, 0x2A3733dBc31980f02b12135C809b5da33BF3a1e9]
    ```

    3 currencies are registered in the `FeeCurrencyDirectory` ✅ As expected, the registered currencies are:

    cUSD ✅
    
    ```sh
    cast call \
    0xe6774BE4E5f97dB10cAFB4c00C74cFbdCDc434D9 \
    "name()" \
    --rpc-url=http://127.0.0.1:8546
    0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000b43656c6f20446f6c6c6172000000000000000000000000000000000000000000
    ```
    
    ```sh
    cast abi-decode \
    "name()(string memory)" \
    "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000b43656c6f20446f6c6c6172000000000000000000000000000000000000000000"
    "Celo Dollar"
    ```
    
    cEUR ✅
    
    ```sh
    cast call \
    0xb7a33b4ad2B1f6b0a944232F5c71798d27Ad9272 \
    "name()" \
    --rpc-url=http://127.0.0.1:8546
    0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000943656c6f204575726f0000000000000000000000000000000000000000000000
    ```
    
    ```sh
    cast abi-decode \
    "name()(string memory)" \
    "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000943656c6f204575726f0000000000000000000000000000000000000000000000"
    "Celo Euro"
    ```
    
    cREAL ✅
    
    ```sh
    cast call \
    0x2A3733dBc31980f02b12135C809b5da33BF3a1e9 \
    "name()" \
    --rpc-url=http://127.0.0.1:8546
    0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001343656c6f204272617a696c69616e205265616c00000000000000000000000000
    ```
    
    ```sh
    cast abi-decode \
    "name()(string memory)" \
    "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001343656c6f204272617a696c69616e205265616c00000000000000000000000000"
    "Celo Brazilian Real"
    ```

1. Get exchange rate for a currency:

    ```sh
    $ cast call \
    0x42Fe5a2A61ed9705eb2F08a04A58CEB606D22f6a \
    "getExchangeRate(address)(uint256, uint256)" \
    0xb7a33b4ad2B1f6b0a944232F5c71798d27Ad9272 \
    --rpc-url=http://127.0.0.1:8546
    1000000000000000000000000 [1e24]
    1000000000000000000000000 [1e24]
    ```
    
    Calling the `getExchangeRate` function on a registered currency returns an exchange rate ✅

1. Terminate Anvil:

    ```sh
    $ kill $(lsof -i tcp:8546 | tail -n 1 | awk '{print $2}')
    ```
    
    Source: [start_anvil.sh](https://github.com/celo-org/celo-monorepo/blob/a36eae5bc2c9c2071fb942199b1ba265f3937c17/packages/protocol/migrations_sol/start_anvil.sh#L9)

</details>


### Related issues

- Fixes #10991

### Backwards compatibility

Yes, only adds to the migration script, doesn't alter existing expectations and behaviour.

### Documentation

No docs changes

arthurgousset committed:
refactor(migrations/start_anvil): renames variable to fix typo

arthurgousset committed:
docs(protocol/create_and_migrate): adds comment for context

arthurgousset committed:
docs(protocol/migrations): adds TODOs and comments for context

arthurgousset committed:
chore(protocol/migrations): adds TODO comment

arthurgousset committed:
feat(protocol/migrations): adds `FeeCurrencyDirectory` migration

Not checked that the current changes are sufficient and working.

arthurgousset committed:
feat(protocol/migrations): sets CurrencyConfig for stabletoken

arthurgousset commented:
Notes-to-self

`create_and_migrate_anvil_devchain.sh`
Source: packages/protocol/migrations_sol/create_and_migrate_anvil_devchain.sh
1. Entry point of workflow
2. Starts anvil with start_anvil.sh script
3. Deploys precompiles with deploy_precompiles.sh script
4. Sets registry contract at specified addresses using registry bytecode

`start_anvil.sh`
Source: packages/protocol/migrations_sol/start_anvil.sh
1. Start anvil instance at specified port and console logs when anvil is running

`deploy_precompiles.sh`
Source: packages/protocol/migrations_sol/deploy_precompiles.sh
1. Gets bytecodes from build artifacts, chooses precompile addresses, and sets bytecode at specified addresses on anvil

`Migration.s.sol`
Source: packages/protocol/migrations_sol/Migration.s.sol
1. run() function is the main entry point of the script

`migrationsConfig.json`
Source: packages/protocol/migrations_sol/migrationsConfig.json
1. Contains contract initialisation data and configs

`FeeCurrencyDirectory.sol`
Source: packages/protocol/contracts-0.8/common/FeeCurrencyDirectory.sol
1. Set a currency config

```sol
/**
   * @notice Sets the currency configuration for a token.
   * @dev This action can only be performed by the contract owner.
   * @param token The token address.
   * @param oracle The oracle address for price fetching.
   * @param intrinsicGas The intrinsic gas value for transactions.
   */
  function setCurrencyConfig(
    address token,
    address oracle,
    uint256 intrinsicGas
  )
```

`IFeeCurrencyDirectory.sol`
Source: packages/protocol/contracts-0.8/common/interfaces/IFeeCurrencyDirectory.sol
1. Specifies interface including CurrencyConfig struct

`run_integration_tests_in_anvil.sh`
Source: packages/protocol/migrations_sol/run_integration_tests_in_anvil.sh
1. Generates devchain and runs simple integration test defined in Integration.t.sol

`integration_tests.sh`
Source: packages/protocol/migrations_sol/integration_tests.sh
1. Uses forge to run test defined in Integration.t.sol script

`Integration.t.sol`
Source: packages/protocol/test-sol/integration/Integration.t.sol
1. High-level test file that defines account1, account2, and registry address
2. Currently doesn't test anything specific to other contracts as far as I can tell

`FeeCurrencyDirectory.t.sol`
Source: packages/protocol/test-sol/common/FeeCurrencyDirectory.t.sol
1. Foundry tests for FeeCurrencyDirectory

`protocol_tests.yml`
Source: .github/workflows/protocol_tests.yml
1. Workflow that runs Foundry-based tests
3. Runs migration test defined in run_integration_tests_in_anvil.sh script

`celo-monorepo.yml`
Source: .github/workflows/celo-monorepo.yml
1. Not relevant to this PR, because it uses the devchain, which is currently Ganache-based.

arthurgousset commented:
## Test migration locally

Requirement:
1. Have foundry installed
2. Be in the `packages/protocol` directory

Steps to reproduce:
1. Compile contracts 
    ```sh
    $ forge --version && forge compile
    ```
2. Run migration test 
    ```sh
    $ ./migrations_sol/run_integration_tests_in_anvil.sh
    ```

## Test `FeeCurrencyDirectory` migration locally

Possible (short-term) solutions:
1. Manually run migration and run the devchain locally
2. Manually call `getCurrencyConfig(stabletoken)` on the local devchain

Possible (longer-term) solutions: 
1. Add `FeeCurrencyDirectory`-related test to `IntegrationTest` (in `Integration.t.sol`), run integration test 
1. Run `FeeCurrencyDirectory` tests (in `FeeCurrencyDirectory.t.sol`) using forge, if they run against the devchain. I don’t think that’s the case, but let’s see.

arthurgousset committed:
Merge branch 'master' into arthurgousset/feat/add-FeeCurrencyDirectory-to-Anvil-migrations

arthurgousset committed:
feat(protocol/migrations_sol): sets example CurrencyConfig

1. adds test oracle address
2. adds test intrinsic gas value

arthurgousset committed:
Merge branch 'master' into arthurgousset/feat/add-FeeCurrencyDirectory-to-Anvil-migrations

arthurgousset commented on packages/protocol/migrations_sol/Migration.s.sol:
```diff
@@ -257,6 +267,15 @@ contract Migration is Script, UsingRegistry {
   }
 
+  // TODO(Arthur): Add FeeCurrencyDirectory function
+  function migrateFeeCurrencyDirectory() public {
```
As of https://github.com/celo-org/celo-monorepo/pull/10992/commits/99b59afbee3990557a54fcc1ed2539a70fd41678, the script (`run_integration_tests_in_anvil.sh`) fails with the following error:

```sh
$ forge --version && forge compile
$ ./migrations_sol/run_integration_tests_in_anvil.sh

# ...

Error:
Compiler run failed:
Error (9582): Member "initialize" not found or not visible after argument-dependent lookup in type(contract IFeeCurrencyDirectory).
   --> migrations_sol/Migration.s.sol:276:30:
    |
276 |       abi.encodeWithSelector(IFeeCurrencyDirectory.initialize.selector)
    |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

It originates in this function I wrote: 

https://github.com/celo-org/celo-monorepo/blob/35258dcd71a2f9036b5dde341dafae7407502f9e/packages/protocol/migrations_sol/Migration.s.sol#L271-L277

arthurgousset replied:
The error makes sense, because there is no `initialize` function in `IFeeCurrencyDirectory.sol`, unlike `IFeeCurrencyWhitelist.sol`:

https://github.com/celo-org/celo-monorepo/blob/35258dcd71a2f9036b5dde341dafae7407502f9e/packages/protocol/contracts/common/interfaces/IFeeCurrencyWhitelist.sol#L5-L5

Instead, it's defined in `FeeCurrencyDirectory.sol`:

https://github.com/celo-org/celo-monorepo/blob/35258dcd71a2f9036b5dde341dafae7407502f9e/packages/protocol/contracts-0.8/common/FeeCurrencyDirectory.sol#L18-L20

arthurgousset replied:
I was calling `initialize()` on  `IFeeCurrencyDirectory`, but that function is defined in `FeeCurrencyDirectory`.
This is fixed in https://github.com/celo-org/celo-monorepo/pull/10992/commits/a36eae5bc2c9c2071fb942199b1ba265f3937c17.

arthurgousset commented:
(note-to-self)
Useful related PRs:
- https://github.com/celo-org/celo-monorepo/pull/10993
- https://github.com/celo-org/celo-monorepo/pull/10875

arthurgousset committed:
chore(protocol/migrations_sol): now uses `FeeCurrencyDirectory`, not interface

I was calling `initialize()`, but that function is defined in `FeeCurrencyDirectory` not `IFeeCurrencyDirectory`.

Also updates string length, which was previously forgotten.

At this point the compilation succeeds as expected, and the migration transactions are successful.

arthurgousset commented:
At this point the compilation succeeds as expected, and the migration transactions are successful.

Next, I'll run a local Anvil instance (with all the migrations), and I'll interact with the deployed `FeeCurrencyDirectory` via `cast` to assert the `FeeCurrencyDirectory` contract works as expected:
1. get `FeeCurrencyDirectory` contract from the `Registry`
2. get the `CurrencyConfig` for the registered mock token

arthurgousset commented on packages/protocol/migrations_sol/Migration.s.sol:
## Local test 

Running Anvil with migrations:

```sh
$ ./migrations_sol/create_and_migrate_anvil_devchain.sh
```

Get `Reserve` contract from the Registry, to assert that Anvil migrations worked and `cast call` is formatted correctly:

```sh
$ export REGISTRY_ADDRESS="0x000000000000000000000000000000000000ce10"
$ export ANVIL_PORT=8546
$ cast call \
$REGISTRY_ADDRESS \
"getAddressForStringOrDie(string calldata identifier)" \
"Reserve" \
--rpc-url='https://alfajores-forno.celo-testnet.org'
0x000000000000000000000000a7ed835288aa4524bb6c73dd23c0bf4315d9fe3e
```

The same call for `FeeCurrencyDirectory` fails:

```sh
$ export REGISTRY_ADDRESS="0x000000000000000000000000000000000000ce10"
$ export ANVIL_PORT=8546
$ cast call \
$REGISTRY_ADDRESS \
"getAddressForStringOrDie(string calldata identifier)" \
"FeeCurrencyDirectory" \
--rpc-url='https://alfajores-forno.celo-testnet.org'
Error:
(code: 3, message: execution reverted: identifier has no registry entry, data: Some(String("0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000206964656e74696669657220686173206e6f20726567697374727920656e747279")))

arthurgousset replied:
My bad, I used the Alfajores Forno RPC above, and not the local Anvil RPC. 

With the correct RPC URL it works ✅

```sh
export REGISTRY_ADDRESS="0x000000000000000000000000000000000000ce10"
export ANVIL_PORT=8546 
cast call \
$REGISTRY_ADDRESS \
"getAddressForStringOrDie(string calldata identifier)" \
"FeeCurrencyDirectory" \
--rpc-url=http://127.0.0.1:$ANVIL_PORT

0x00000000000000000000000042fe5a2a61ed9705eb2f08a04a58ceb606d22f6a
```

```sh
cast abi-decode \
"getAddressForStringOrDie(string calldata identifier)(address)" \
"0x00000000000000000000000042fe5a2a61ed9705eb2f08a04a58ceb606d22f6a"

0x42Fe5a2A61ed9705eb2F08a04A58CEB606D22f6a
```

arthurgousset commented:
(note-to-self)

1. Check that Anvil is running

    ```sh
    $ export ANVIL_PORT=8546
    $ nc -z localhost $ANVIL_PORT
    Connection to localhost port 8546 [tcp/*] succeeded!
    ```

    Source: [start_anvil.sh](https://github.com/celo-org/celo-monorepo/blob/a36eae5bc2c9c2071fb942199b1ba265f3937c17/packages/protocol/migrations_sol/start_anvil.sh#L7)

2. Terminate Anvil:

    ```sh
    $ export ANVIL_PORT=8546
    $ kill $(lsof -i tcp:$ANVIL_PORT | tail -n 1 | awk '{print $2}')
    ```
    
    Source: [start_anvil.sh](https://github.com/celo-org/celo-monorepo/blob/a36eae5bc2c9c2071fb942199b1ba265f3937c17/packages/protocol/migrations_sol/start_anvil.sh#L9)

arthurgousset commented:
At this point (https://github.com/celo-org/celo-monorepo/pull/10992/commits/a36eae5bc2c9c2071fb942199b1ba265f3937c17), the local tests pass.

### Local test

We can:

1. get the `FeeCurrencyDirectory` address from the `Registry`:

    ```sh
    export REGISTRY_ADDRESS="0x000000000000000000000000000000000000ce10"
    export ANVIL_PORT=8546 
    cast call \
    $REGISTRY_ADDRESS \
    "getAddressForStringOrDie(string calldata identifier)" \
    "FeeCurrencyDirectory" \
    --rpc-url=http://127.0.0.1:$ANVIL_PORT
    
    0x00000000000000000000000042fe5a2a61ed9705eb2f08a04a58ceb606d22f6a
    ```
    
    ```sh
    cast abi-decode \
    "getAddressForStringOrDie(string calldata identifier)(address)" \
    "0x00000000000000000000000042fe5a2a61ed9705eb2f08a04a58ceb606d22f6a"
    
    0x42Fe5a2A61ed9705eb2F08a04A58CEB606D22f6a
    ```

1. get registered currencies from the `FeeCurrencyDirectory` contract

    ```sh
    export FEECURRENCYDIRECTORY_PROXY_ADDRESS=0x42Fe5a2A61ed9705eb2F08a04A58CEB606D22f6a
    export ANVIL_PORT=8546
    export MOCKTOKEN_ADDRESS=0xefB84935239dAcdecF7c5bA76d8dE40b077B7b33
    
    cast call \
    $FEECURRENCYDIRECTORY_PROXY_ADDRESS \
    "getCurrencies()" \
    --rpc-url=http://127.0.0.1:$ANVIL_PORT
    
    0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003000000000000000000000000e6774be4e5f97db10cafb4c00c74cfbdcdc434d9000000000000000000000000b7a33b4ad2b1f6b0a944232f5c71798d27ad92720000000000000000000000002a3733dbc31980f02b12135c809b5da33bf3a1e9
    ```
    
    ```sh
    cast abi-decode \
    "getCurrencies()(address[] memory)" \
    "0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003000000000000000000000000e6774be4e5f97db10cafb4c00c74cfbdcdc434d9000000000000000000000000b7a33b4ad2b1f6b0a944232f5c71798d27ad92720000000000000000000000002a3733dbc31980f02b12135c809b5da33bf3a1e9"
    
    [0xe6774BE4E5f97dB10cAFB4c00C74cFbdCDc434D9, 0xb7a33b4ad2B1f6b0a944232F5c71798d27Ad9272, 0x2A3733dBc31980f02b12135C809b5da33BF3a1e9]
    ```

    As expected, the registered currencies are 

    cUSD:
    
    ```sh
    cast call \
    0xe6774BE4E5f97dB10cAFB4c00C74cFbdCDc434D9 \
    "name()" \
    --rpc-url=http://127.0.0.1:8546
    0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000b43656c6f20446f6c6c6172000000000000000000000000000000000000000000
    ```
    
    ```sh
    cast abi-decode \
    "name()(string memory)" \
    "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000b43656c6f20446f6c6c6172000000000000000000000000000000000000000000"
    "Celo Dollar"
    ```
    
    cEUR:
    
    ```sh
    cast call \
    0xb7a33b4ad2B1f6b0a944232F5c71798d27Ad9272 \
    "name()" \
    --rpc-url=http://127.0.0.1:8546
    0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000943656c6f204575726f0000000000000000000000000000000000000000000000
    ```
    
    ```sh
    cast abi-decode \
    "name()(string memory)" \
    "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000943656c6f204575726f0000000000000000000000000000000000000000000000"
    "Celo Euro"
    ```
    
    cREAL:
    
    ```sh
    cast call \
    0x2A3733dBc31980f02b12135C809b5da33BF3a1e9 \
    "name()" \
    --rpc-url=http://127.0.0.1:8546
    0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001343656c6f204272617a696c69616e205265616c00000000000000000000000000
    ```
    
    ```sh
    cast abi-decode \
    "name()(string memory)" \
    "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001343656c6f204272617a696c69616e205265616c00000000000000000000000000"
    "Celo Brazilian Real"
    ```

arthurgousset committed:
chore(protocol/migrations_sol): adds `MockOracle` to `CurrencyConfig`s

This means, all currencies registered in the `FeeCurrencyDirectory` now support the `getExchangeRate` function, which uses the MockOracle contract under the hood.

arthurgousset commented:
As of 8b910b7b3d9b988f6f20b02d5f5959c618682cc1, the migration script sets the oracle in `CurrencyConfig` to a [`MockOracle.sol`](https://github.com/celo-org/celo-monorepo/blob/7828cd46299b1ef214436609c8b41c547adf83e4/packages/protocol/contracts-0.8/common/mocks/MockOracle.sol) contract (which implements [`IOracle.sol`](https://github.com/celo-org/celo-monorepo/blob/43a3146c068de25cda53b0aff4853ea4bc37e887/packages/protocol/contracts-0.8/common/interfaces/IOracle.sol)), and sets an arbitrary exchange rate of $\frac{10}{5}$.

This works as expected :white_check_mark:

Calling the `getExchangeRate` function on a registered currency returns the arbitrary exchange rate

```sh
$ cast call \
0x42Fe5a2A61ed9705eb2F08a04A58CEB606D22f6a \
"getExchangeRate(address)(uint256, uint256)" \
0xb7a33b4ad2B1f6b0a944232F5c71798d27Ad9272 \
--rpc-url=http://127.0.0.1:8546/
10
5
```

arthurgousset committed:
Merge branch 'master' into arthurgousset/feat/add-FeeCurrencyDirectory-to-Anvil-migrations

pahor167 commented on packages/protocol/migrations_sol/Migration.s.sol:
```diff
@@ -11,6 +11,9 @@ import "@celo-contracts/common/interfaces/IProxy.sol";
 import "@celo-contracts/common/interfaces/IRegistry.sol";
 import "@celo-contracts/common/interfaces/IFreezer.sol";
 import "@celo-contracts/common/interfaces/IFeeCurrencyWhitelist.sol";
+// TODO(Arthur): import FeeCurrencyDirectory interface?
```
Do we need this todo ?

arthurgousset replied:
No, I removed various `TODO` comments I left for myself in this commit https://github.com/celo-org/celo-monorepo/pull/10992/commits/18eb1638618457022c443c6571968eb47dbe9848

pahor167 commented on packages/protocol/migrations_sol/Migration.s.sol:
```diff
@@ -198,6 +209,8 @@ contract Migration is Script, UsingRegistry {
 
     migrateFreezer();
     migrateFeeCurrencyWhitelist();
+    // TODO(Arthur): Add FeeCurrencyDirectory here
```
Do we need this todo ?

pahor167 replied:
Can you please check whether all todos are needed ?

arthurgousset replied:
> Do we need this todo ?

No, I removed various TODO comments I left for myself in this commit https://github.com/celo-org/celo-monorepo/commit/18eb1638618457022c443c6571968eb47dbe9848

arthurgousset replied:
> Can you please check whether all todos are needed ?

Yes, that's my bad, the PR is not ready for review. I need to remove a couple `TODO` comments I left for myself, and add a PR description.

pahor167 commented on packages/protocol/migrations_sol/Migration.s.sol:
```diff
@@ -404,7 +426,32 @@ contract Migration is Script, UsingRegistry {
+    This is an arbitrary hex address I took from celoscan.
+    For now, I only want to test if I can set a currency config.
+    */
+    MockOracle mockOracle = new MockOracle();
```
Can we use SortedOracles instead ?

arthurgousset replied:
Yes, it looks like the latest [`SortedOracles.sol`](https://github.com/celo-org/celo-monorepo/blob/a4c608b4211794bb1651386785d2661b9679e8b7/packages/protocol/contracts/stability/SortedOracles.sol) implementation has a `getExchangeRate()` implementation:

https://github.com/celo-org/celo-monorepo/blob/ed539fda163b1f5d7ce6a2b40944891c4c3a9ad2/packages/protocol/contracts/stability/SortedOracles.sol#L329-L333

Last I looked, [`SortedOracles.sol`](https://github.com/celo-org/celo-monorepo/blob/a4c608b4211794bb1651386785d2661b9679e8b7/packages/protocol/contracts/stability/SortedOracles.sol) did not implement the [`IOracle.sol`](https://github.com/celo-org/celo-monorepo/blob/43a3146c068de25cda53b0aff4853ea4bc37e887/packages/protocol/contracts-0.8/common/interfaces/IOracle.sol) interface. 

Nit: Looks like the class definition does not specify that "`SortedOracles is IOracle`". Any reason we didn't add this? Happy to make this small change if it makes sense.

https://github.com/celo-org/celo-monorepo/blob/ed539fda163b1f5d7ce6a2b40944891c4c3a9ad2/packages/protocol/contracts/stability/SortedOracles.sol#L39-L39

pahor167 replied:
This is a miss. If you can, please add IOracle to SortedOracles

arthurgousset replied:
https://github.com/celo-org/celo-monorepo/pull/10992/commits/1c04e8ba7ea1c632754709bd560f7b9c58f0aa90 replaces `MockOracle` with `SortedOracles` ✅

https://github.com/celo-org/celo-monorepo/pull/10992/commits/78e6cf273e87f2445bed4912f99486bce93f5b08 adds `IOracle` to the class definition ✅

arthurgousset replied:
### Test

> As of [1c04e8b](https://github.com/celo-org/celo-monorepo/commit/1c04e8ba7ea1c632754709bd560f7b9c58f0aa90), the migration script sets the oracle in `CurrencyConfig` to the [`SortedOracles.sol`](https://github.com/celo-org/celo-monorepo/blob/a4c608b4211794bb1651386785d2661b9679e8b7/packages/protocol/contracts/stability/SortedOracles.sol) contract (which implements [IOracle.sol](https://github.com/celo-org/celo-monorepo/blob/43a3146c068de25cda53b0aff4853ea4bc37e887/packages/protocol/contracts-0.8/common/interfaces/IOracle.sol)).
> 
> This works as expected ✅
> 
> Calling the `getExchangeRate` function on a registered currency returns an exchange rate:
> 
> ```shell
> $ cast call \
> 0x42Fe5a2A61ed9705eb2F08a04A58CEB606D22f6a \
> "getExchangeRate(address)(uint256, uint256)" \
> 0xb7a33b4ad2B1f6b0a944232F5c71798d27Ad9272 \
> --rpc-url=http://127.0.0.1:8546
> 1000000000000000000000000 [1e24]
> 1000000000000000000000000 [1e24]
> ```

Source: https://github.com/celo-org/celo-monorepo/pull/10992#issuecomment-2120486484

arthurgousset committed:
chore(protocol/migrations_sol): removes `TODO` comments

arthurgousset committed:
Merge branch 'arthurgousset/feat/add-FeeCurrencyDirectory-to-Anvil-migrations' of github.com:celo-org/celo-monorepo into arthurgousset/feat/add-FeeCurrencyDirectory-to-Anvil-migrations

arthurgousset committed:
chore(protocol:SortedOracles.sol): Adds `IOracle` to contract definition

The `SortedOracles.sol` contract already implements the `IOracle.sol` interface. This commit only adds it to the contract definition for ease of reference.

arthurgousset committed:
chore(protocol/migrations_sol): Replaces `MockOracle` with `SortedOracles`

The latest `SortedOracles.sol` implementation implements the `IOracle.sol` interface. That means it implements the required `getExchangeRate()` function.

That means the `MockOracle.sol` contract is not required in the migrations.

arthurgousset commented:
As of https://github.com/celo-org/celo-monorepo/pull/10992/commits/1c04e8ba7ea1c632754709bd560f7b9c58f0aa90, the migration script sets the oracle in `CurrencyConfig` to the [`SortedOracles.sol`](https://github.com/celo-org/celo-monorepo/blob/a4c608b4211794bb1651386785d2661b9679e8b7/packages/protocol/contracts/stability/SortedOracles.sol) contract (which implements [IOracle.sol](https://github.com/celo-org/celo-monorepo/blob/43a3146c068de25cda53b0aff4853ea4bc37e887/packages/protocol/contracts-0.8/common/interfaces/IOracle.sol)).

This works as expected ✅

Calling the `getExchangeRate` function on a registered currency returns an exchange rate:

```sh
$ cast call \
0x42Fe5a2A61ed9705eb2F08a04A58CEB606D22f6a \
"getExchangeRate(address)(uint256, uint256)" \
0xb7a33b4ad2B1f6b0a944232F5c71798d27Ad9272 \
--rpc-url=http://127.0.0.1:8546
1000000000000000000000000 [1e24]
1000000000000000000000000 [1e24]
```

arthurgousset committed:
chore(protocol/migrations_sol): removes unused and duplicate imports

arthurgousset committed:
style(protocol/migrations_sol): linting

arthurgousset committed:
test(protocol/test-sol/Validators): updates L2 error messages

Now matches error message defined in `isL2Check.sol`
Source: https://github.com/celo-org/celo-monorepo/blob/668a4050f7ee067948498c333da3feec85fa03fe/packages/protocol/contracts-0.8/common/IsL2Check.sol#L32-L33
